src/Tools/dotnet-monitor/CollectionRules/CollectionRuleService.cs

```diff
@@ -0,0 +1,178 @@
+
+            foreach (string ruleName in ruleNames)
+            {
+                TaskCompletionSource<object> startedSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                startedSources.Add(startedSource);
+
+                _runTasks.Add(RunRuleAsync(
+                    _actionListExecutor,
+                    _triggerOperations,
+                    _optionsMonitor,
+                    endpointInfo,
+                    ruleName,
+                    startedSource,
+                    linkedSource.Token));
+            }
+
+            await Task.WhenAll(startedSources.Select(s => s.Task).ToArray());
+
+            KeyValueLogScope scope = new();
+            scope.AddCollectionRuleEndpointInfo(endpointInfo);
+            using IDisposable loggerScope = _logger.BeginScope(scope);
+
+            _logger.AllCollectionRulesStarted();
+        }
+
+        private async Task RunRuleAsync(
+            ActionListExecutor actionListExecutor,
+            ICollectionRuleTriggerOperations triggerOperations,
+            IOptionsMonitor<CollectionRuleOptions> optionsMonitor,
+            IEndpointInfo endpointInfo,
+            string ruleName,
+            TaskCompletionSource<object> startedSource,
+            CancellationToken token)
+        {
+            KeyValueLogScope scope = new();
+            scope.AddCollectionRuleEndpointInfo(endpointInfo);
+            scope.AddCollectionRuleName(ruleName);
+            using IDisposable loggerScope = _logger.BeginScope(scope);
+
+            using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(
+                _disposalTokenSource.Token,
+                token);
+
+            try
+            {
+                CollectionRuleOptions options = optionsMonitor.Get(ruleName);
+
+                if (null != options.Filters)
+                {
+                    DiagProcessFilter filter = DiagProcessFilter.FromConfiguration(options.Filters);
+                    // TODO: Filter collection rules by process information; this requires pushing
+                    // more of the process information into IEndpointInfo. IProcessInfo is only
+                    // available through the IDiagnosticServices implementation, which is created from
+                    // the entries in the IEndpointInfoSource implementation. The collection rules
+                    // are started before the process is registered with the IEndpointInfoSource.
+                }
+
+                _logger.CollectionRuleStarted(ruleName);
+
+                CollectionRuleContext context = new(ruleName, options, endpointInfo, _logger);
+
+                await using CollectionRulePipeline pipeline = new(
+                    actionListExecutor,
+                    triggerOperations,
+                    context);
+
+                await pipeline.StartAsync(linkedSource.Token);
```

Comment:
I'm a little worried about overlapping actions from independent rules. For example, if one rule triggers an action to collect logs for a minute, that action is likely to overlap with another rule/action. Serializing or coalescing actions might make sense.